### DISCUSSION TITLE
Script Command Overhaul

### DISCUSSION THREAD

**Discussion by OrionSR** [sannybuilder/dev#75] [Showroom]
### **[GTA3CommandOverhaul](https://docs.google.com/spreadsheets/d/1lN7lDTrjISqF-myuoJjyWLjD8BUyfUCEEA8EyGU97MY/edit?usp=sharing)** - Google Sheet (wip)

After several long days I've only managed to work my way through a third of the GTA3 opcode examples, but I've made enough progress to give a good impression of what I'm trying to create.  I was hoping that starting with fastman92's script command definitions would give me a better head start because I've already worked through the common SA commands, but the GTA3 definitions are still pretty messed up. This is going to take a while.

Please comment on the conventions being used. Issues with keyword inconsistencies and formatting will likely work themselves out as the file is reprocessed through an updated SCM.ini. Is it worth continuing in this direction? 

Please keep in mind that this version is not intended to be the final product. Many of the conventions are in place to help me reprocess consistent keywords (True/False), INI updates (handles for classes), and enum documentation.

I need to formalize my conventions as I continue to work so I'll add more specific info shortly.

**Comment by OrionSR** [comment]
### **Variables**

I'm working towards a more descriptive set of commands that are independent of the specific main.scm files. The current goal is to create a compiled example.scm that includes all opcodes so the the make_opcode tool can generate a complete list of examples with an updated command set. The compiled data can be disassembled using INI files with any parameter order.

I'm trying to avoid local variables since they aren't descriptive, and not using arrays since they clutter the examples unnecessarily. 

**True/False** - Since Sanny doesn't support a Boolean data  type I'll remove the $ from $True and $False to denote the nature of the setting. Descriptors for the data should imply the proper value without additional reference. For example, `for_daytime False` should obviously apply to night, and I can predict what will happen when I set `show_sphere True' (will fix later). 

**$Enum** -  Actual data would be preferred but I want to keep track of where additional documentation is needed. Descriptors for enums include a trailing underscore because that currently works well for auto-completing constants.

**Integers** 
- $Integer and $Integer2 are used in math operations that require a global variables and for return values of integer type.
- 0@ and @1 are used to denote integer values when a local variable must be used
- $Hours, $Minutes
- $Timer, $Counter - odd opcodes that require a var_int that supplies a value and then holds a handle, I think.
- $Global_Var - used like $OnMission, links a global variable to the game somehow.
- $Bit_Flags

**Floats**
- $Float and $Float2 for generic use in math operations and return values.
- 2@ and 3@ for math operations.
- $X, $Y, $Z and $X2, $Y2 $Z2 (etc?) significantly reduces the length of coordinate commands.
- $Degrees

**Handles**
- $Handle - used to hold return values for any type of handle. 
- $Actor_Handle, $Actor_Handle2 - used in actor specific handles
- $Audio_Handle, $Blip_Handle, $Car_Handle, $Cargen_Handle, $Object_Handle, $Pickup_Handle
- $Garage_Handle, $Fire_Handle


Updating as I progress, but that's it for now.

**Comment by OrionSR** [comment]
### **Enumerations**

As I've been working my way through the scripts I've been marking descriptors of fields that might be appropriate for enums with a trailing underscore. Often I'll find documentation that could be adapted to enums, but then losing track of the link I plan to use this post to update the enums I've identified with links to the documentation. This should help also help with Sanny's overall documentation. It is not always clear of the values given are appropriate for GTA3.

'AUDIO' [gtamods](https://gtamods.com/wiki/Audio_directory_(III))
'TXD' [gtamods](https://gtamods.com/wiki/Txd_(folder))
Import Slot ([gtamods](https://gtamods.com/wiki/03D4)) Tough to enumerate as the list varies depending on the garage
Chase [gtamods](https://gtamods.com/wiki/Chase.dat) No enum, don't get it. Doesn't seem useful.

pedtype_ [gtamods](https://gtamods.com/wiki/Ped_type)
driving_style_ [gtamods](https://gtamods.com/wiki/00AE)
car_mission_ [gtagmodding](https://gtagmodding.com/opcode-database/opcode/00AF/)
print_style_[gtamods]( https://gtamods.com/wiki/00BA)
button_ [SA gtagmodding](https://gtagmodding.com/sanandreas/pads/)[ gtamods](https://gtamods.com/wiki/00E1)
weapon_ [gtamods](https://gtamods.com/wiki/Weapon#GTA_III)
lock_mode_ [gtagmodding](https://gtagmodding.com/opcode-database/opcode/020A/)
car_color_  carcols.dat
camera_mode_ [gtagmodding](https://gtagmodding.com/gta3/camera-modes/)
switchstyle_ [gtagmodding](https://gtagmodding.com/opcode-database/opcode/0158/) 
blip_color_ [gtamods](https://gtamods.com/wiki/0165#Blip_colors)
visibility_ inconsistent, maybe same as radar_mode
radar_mode_ [gtamods](https://gtamods.com/wiki/018B)
route_mode_ [gtamods](https://gtamods.com/wiki/01E1)
pickup_type_ [gtamods](https://gtamods.com/wiki/0213)
rgb_ [w3schools](https://www.w3schools.com/colors/colors_hex.asp) [rapidtables](https://www.rapidtables.com/web/color/RGB_Color.html)
threat_ (**Bit Flags**) [gtamods](https://gtamods.com/wiki/03F1#Ped_threats)
garage_type_ [gtamods](https://gtamods.com/wiki/Garage#GTA_III)
gang_ [gtamods](https://gtamods.com/wiki/Gang#GTA_III)
bomb_type_ [gtamods](https://gtamods.com/wiki/0228#Car_bomb_stages)
personality_ [gtamods](https://gtamods.com/wiki/0243) (ped_stats)
group_anim_ [gtamods](https://gtamods.com/wiki/0245)
wait_anim_ [gtamods](https://gtamods.com/wiki/0245)
corona_type_ [gtamods](https://gtamods.com/wiki/024F#Types_of_Coronas)
lensflare_ [gtamods](https://gtamods.com/wiki/024F#Types_of_Lens_Flare)
particle_ [gtamods](https://gtamods.com/wiki/Particle.cfg#List_of_particles_in_III)
particle_effect_ [gtamods](https://gtamods.com/wiki/039D)
sprite_blip_ [gtamods](https://gtamods.com/wiki/Blip#GTA_III)
ped_group_ [gtamods](https://gtamods.com/wiki/0324) for VC, need to reconcile with pedgrp.dat
font_ [gtamods](https://gtamods.com/wiki/0349)
explosion_ [gtagmodding (sa)](https://gtagmodding.com/sanandreas/explosions/) 
motion_blur_ [gtamodding](https://gtamods.com/wiki/0374)
island_ [gtamods](https://gtamods.com/wiki/ZONE#GTA_III) (1,2,3) See also: [041F:](https://gtamods.com/wiki/041F)
level_ [gtamods](https://gtamods.com/wiki/ZONE#GTA_III) (0,1,2)
hud_object_ [gtamods](https://gtamods.com/wiki/03E7) 
status_ [gtamods](https://gtamods.com/wiki/SCM_language_III/VC_definitions#STATUS)
door_ [gtamods](https://gtamods.com/wiki/SCM_language_III/VC_definitions#DOOR)
mission_index_ [gtamods](https://gtamods.com/wiki/0417)
say_ [gtamods](https://gtamods.com/wiki/041C)
stat_ [gtamods](https://gtamods.com/wiki/List_of_statistics_(III))
seat_ [gtamods](https://gtamods.com/wiki/0431)
facing_ [gtamods](https://gtamods.com/wiki/0464)
tempact_ [gtamods](https://gtamods.com/wiki/0477) temp action
pad_ [gtamods](https://gtamods.com/wiki/00E1)
timer_ [gtamods](https://gtamods.com/wiki/SCM_language_III/VC_definitions#TIMER) count_ direction_ ?

**Comment by OrionSR** [comment]
### Opcode Overhaul

I have finished my first pass through the example codes (opcode.txt) and starting the process of updating the opcode definitions (SCM.ini). The process should help shake out more errors and typos in the manual updates of the examples. My current goal is to simply get the examples to compile - adding missing opcodes, and resolving issues with parameter count in the examples. Once compiled I can decompile, compare the output, and make adjustments until everything matches. 

_I need to write this next bit down as reference as I work._

[SB's Opcode Parameters](https://docs.sannybuilder.com/edit-modes/opcodes-list-scm.ini#opcode-parameters)
d% = anything
p% = label pointer
o% = object models all types
o% = objects
m% = ped & car models (see default.dat, if < 200)
g% = gxt reference
x% = external script
(Does t% still function like o%?)

Like d%, other parameter types have no special formatting. I'm using several for Informational reference of specific data types.
a% = a string within a gxt string (as used)
b% = boolean (as used) but always resolves to True or False
s% = string (as used for other games) here it's purely informational
t% = legacy form of m%, used here as a placeholder until m% is working
h% = handle, integer stored in var or lvar (as used inconsistently)

i% = integer - informational
f% = float - informational
v% = var - when a global variable needs to be used (math, variables linked to exe)
l% = local - when a local variable needs to be used (math)
c% = constant - for constant specific opcodes, maybe
e% = enumeration

Keyword Updates:
`00D6=if` replaces `00D6=andor`
`03E2=register_el_burro_time` replaces `03E2=set_char_obj_leave_any_car`

**Comment by x87** [comment]
@OrionSR I'm thinking of using `b:` placeholder as a backward-compatible reference to the newer type system (specifically enumerated params). This idea exploits the fact that all Sanny versions ignore text after `b:` and handle this type of param as `any` (%d).

In a next version I'd fill this placeholder with enum names so that Sanny becomes aware of which enum type to use to replace numbers with during disassemble and do the opposite conversion during compilation. E.g.:

```
009A=6,%6d% = create_actor_pedtype %1b:pedtype% model %2m% at %3d% %4d% %5d%
```

this would make changes to INI files two-way compatible with any Sanny version.

Old INI files with `b:true/false` definitions would become incompatible with that newer Sanny but I think I can make it recognize the old format with the slash separated values vs the new one.

Let me know what do you think? Does it make sense or just update INI altogether and force people using correct Sanny version with correct INI file?

**Comment by OrionSR** [comment]
Before I forget, the precision of the decompiled floating-point values is terrible - not hard to adjust for the next run through. I'm assuming that GTA3 is using 16-bit "short floats" but I'm having a hard confirming this with available documentation. 

In general, I'm not comfortable with the b-for-enums solution. But I suspect that's because I've been thinking along different lines. I will reconsider as I continue to work. (Slept on it; ideas are starting resolve.)

First of all, this is a critical discussion that will have a large impact on how the command overhaul proceeds. I'm having a hard time visualizing what the final product will look like as the examples progress from RockStyle-SCR to RockStyle-SB with classes without breaking anything. My lack of experience using classes is major problem. 

Please post a few examples of how enums and bools might appear when used within classes.
Can t% have the same functionality as %m? 

I think it would be pretty cool if `%2b:day/night%` could be resolved as expected, but in general I believe this is the wrong way to think about bools. I'm looking for a way to include Deji's documentation of boolean parameters as unique true/false values. However, I'm a bit concerned about restricting the values that can be used. Some of my guesses might be wrong.

So far I've identified about 45 data types that are appropriate for enumeration. Everything I could resolve as boolean was skipped.  Including enum declarations and documentation of the uncounted boolean data types would significantly increase the complexity of the enumeration process. Boolean descriptors are freeform; they can be tweaked for clarification without needing to fix something somewhere else.

Compatibility Between Sanny Versions: In general, I would hope that legacy formats for opcodes (SCM.ini) to be forward compatible with newer versions of SB. I would not expect newer command sets to be backward compatible. However. I think backwards compatibility is a worthy goal if it can be accomplished without too much difficulty. I haven't seen e% used elsewhere, yet, and as a generic data type the fields should resolve to the same values as before... Unless I do something weird like try to define an enum key like the legacy bools, or, maybe, connect the descriptor key directly to e% with a dot separator.

How about, e% for enumeration, and the key is the preceding descriptor. Perhaps require a trailing dot for identification. Currently I'm using trailing underscores so I can simulate enums with constants, but in the examples below I've worked in the dot.  

I'm just playing around here, trying to get a feel for what things might look like:
```
0155=3,set_zone_car_density %1z% for_daytime %2b% density %3d%  // z%?
0155: set_zone_car_density 'ZONE' for_daytime False density 20  // boolean
0155: set_zone_car_density 'ZONE' for_daytime 0 density 20  // compatible

0155=3,set_zone_car_density %1s% %2b:day/night% density %3i%  // false positive for t%, false negative for b%  
0155: set_zone_car_density 'ZONE' night density 20 // how would this compile?

0155=3,set_zone_car_density %1s% timeofday. %2e% density %3i%  // as enum
0155: set_zone_car_density 'ZONE' timeofday. 0 density 20  // old, ambiguous
0155: set_zone_car_density 'ZONE' TimeOfDay.Night density 20  // new

009A=6,create_char pedtype. %1e% model %2t% coord %3f% %fd% %fd% store_to %6h%
009A: create_char pedtype. 4 model #MALE01 coord $X $Y $Z store_to $Actor_Handle
009A: create_char PedType.CivMale model #MALE01 coord $X $Y $Z store_to $Actor_Handle

0229=3,change_car_colour %1h% primary carcols. %2e% secondary carcols. %3e%
0229: change_car_colour $Car_Handle primary carcols. 6 secondary carcols. 7
0229: change_car_colour $Car_Handle primary CarCols.Taxi_Yellow secondary CarCols.Striking_Blue

0229=3,change_car_colour %1h% carcols. %2e% %3e%
0229: change_car_colour $Car_Handle carcols. 6 7
0229: change_car_colour $Car_Handle CarCols.Taxi_Yellow.Striking_Blue
```

**Comment by x87** [comment]
>Before I forget, the precision of the decompiled floating-point values is terrible - not hard to adjust for the next run through. I'm assuming that GTA3 is using 16-bit "short floats" but I'm having a hard confirming this with available documentation.

this is correct. GTA3 uses 16-bit integers to represent decimal values
https://gtamods.com/wiki/Talk:Mission_Scripting_(Overview)#Fixed-point_remark

>Can t% have the same functionality as %m?

yes, I can do that, but in general I'd avoid duplicating things in updated INIs (i.e. INI file should use only %t or only %m).

>How about, e% for enumeration, and the key is the preceding descriptor. 

I feel it makes heavy impact on the code decompiled with older Sanny versions and new INI and adds unnecessary ambigous syntax to opcode descriptors. Nowadays words between parameters are just another form of comments and I would like to keep it, not making them part of the syntax.

If using `b:` syntax seems awkward I think we could use e% type followed an enum as in:

```
0229=3,change_car_colour %1h% primary %2e:carcols% secondary %3e:carcols%
```
which decompiles to

```
0229: change_car_colour $Car_Handle primary CarCols.Taxi_Yellow secondary CarCols.Striking_Blue
```
or
```
$Car_Handle.SetColor(CarCols.Taxi_Yellow,  CarCols.Striking_Blue)
```
with classes enabled

I'm not too worried about new syntax being incompatible with older Sanny version as it would encourage people migrate to newer Sanny sooner to benefit from the updated opcode definitions. 

>Please post a few examples of how enums and bools might appear when used within classes.

Examples using updates classes from #62 
```
Car.setImmunities(0@, True, True, True, True, True)
Actor.asAnimationVB(-1, "LRGIRL_IDLELOOP", "LOWRIDER", 4.0, False, False, False, False, -1)
Actor.CreatePersonnel(2@, CivMale, 7, 0.0, 0.0, 0.0)
```
note that classes do not have a notion of enums so old "extended" params still appear (CivMale). In the future the last command should decompile as 

```
Actor.CreatePersonnel(2@, PedType.CivMale, 7, 0.0, 0.0, 0.0)
```

**Comment by x87** [comment]
Added new enums to classes.db to replace old "extended" params
https://github.com/sannybuilder/data/pull/6/commits/ba3a4ed2712cc3f8c97816f08e3d6450d6c166e7

I chose camelCase naming convention over R* SCREAMING_SNAKE_CASE to better match existing Sanny syntax.
